sdk/daml-script/test/daml/upgrades/DataChanges.daml

```diff
@@ -286,3 +317,336 @@ templateEnumDowngradeFromOld =
+  res <- (actAs a <> packagePreference [dataChangesV1]) `trySubmit` exerciseCmd cidV2 V2.OptionalFieldOnChoiceAddedCall { existingField = "V2", newField = Some "V2" }
+
+  case res of
+    Left (UnknownError msg) | "An optional contract field (\"newField\") with a value of Some may not be dropped during downgrading." `isInfixOf` msg -> pure ()
+    _ -> assertFail $ "Expected specific failure but got " <> show res
+
+{- PACKAGE
+name: data-changes-iface
+versions: 1
+-}
+
+{- MODULE
+package: data-changes-iface
+contents: |
+  module IfaceMod where
+
+  data MyUnit = MyUnit {}
+
+  interface I where
+    viewtype MyUnit
+
+    nonconsuming choice NoOp : ()
+      with
+        ctl : Party
+      controller ctl
+      do
+        assertFail "aaa"
+        pure ()
+-}
+
+{- MODULE
+package: data-changes
+contents: |
+  module OptionalFieldsAddedUsingIface where
+
+  import IfaceMod
+
+  template OptionalFieldsAddedUsingIface
+    with
+      party : Party
+      newField : Optional Text                 -- @V  2
+    where
+      signatory party
+
+      interface instance I for OptionalFieldsAddedUsingIface where
+        view = MyUnit {}
+
+  template IChoiceBodyCall
+    with
+      party : Party
+    where
+      signatory party
+      choice IChoiceBodyCallChoice : ()
+        with
+          cidI : ContractId I
+          ctl : Party
+        controller ctl
+        do
+          exercise cidI (NoOp ctl)
+
+  template ILocalTestDummy
+    with
+      party : Party
+    where
+      signatory party
+      choice ILocalTestDummyChoice : ()
+        controller party
+        do
+          cid <- create
+            (OptionalFieldsAddedUsingIface
+              party
+              (Some "V1") -- @V  2
+            )
+          let cidIface = coerceContractId @OptionalFieldsAddedUsingIface @I cid
+          exercise cidIface (NoOp party)
+-}
+
+dataChangesV1 : PackageId
+dataChangesV1 = getPackageId "data-changes-1.0.0"
+
+templateOptionalFieldsSetWhenDowngradingTest : Show b => Text -> (Party -> Party -> Script (Either SubmitError b)) -> (Text, Test)
+templateOptionalFieldsSetWhenDowngradingTest trailer f = ("Fails if an optional field is set when downgrading (" <> trailer <> ")", s)
+  where
+  s = test $ do
+    a <- allocatePartyOn "alice" participant0
+    b <- allocatePartyOn "bob" participant0
+    res <- f a b
+    case res of
+      Left (DevError Upgrade msg) | "An optional contract field with a value of Some may not be dropped during downgrading" `isInfixOf` msg -> pure ()
+      _ -> assertFail $ "Expected specific failure but got " <> show res
+
+tofswdViaInterfaceInCommandGlobal : (Text, Test)
+tofswdViaInterfaceInCommandGlobal = templateOptionalFieldsSetWhenDowngradingTest "dynamic, global,    in command" $ \a b -> do
+  cid <- a `submit` createCmd V2.OptionalFieldsAddedUsingIface with party = a, newField = Some "V1"
+  let cidIface = coerceContractId @V2.OptionalFieldsAddedUsingIface @Iface.I cid
```

Comment:
Nit: for the special case of coercing to an interface ID, there's a toInterfaceContractId function (which, as far as I understand, does the same thing but statically checks that this is safe, i.e. that the template has an instance for that interface).

Reply:
fixed